File: 12_MAP_ALGEBRA/12_MAP_ALGEBRA.R
library(terra)
library(sf)
library(classInt)
library(tmap)
library(stars)

# Set working directory
workspace <- "C:/Users/alvarece/Desktop/MODELAMIENTO Y PROGRAMACION GIS/PRACTICAS/DATOS/RASTER"

# Load rasters
nir <- rast(file.path(workspace, "LC09_20240831_B5.tif"))
red <- rast(file.path(workspace, "LC09_20240831_B4.tif"))
dem <- rast(file.path(workspace, "DTM.tif"))

# Load shapefile
clip_shape <- st_read(file.path(workspace, "QUITO.shp"))

# Reproject rasters to EPSG:4326 if needed
target_crs <- "EPSG:4326"

if (!crs(nir) == target_crs) nir <- project(nir, target_crs)
if (!crs(red) == target_crs) red <- project(red, target_crs)
if (!crs(dem) == target_crs) dem <- project(dem, target_crs)

# Clip rasters
nir_c <- crop(nir, vect(clip_shape)) |> mask(vect(clip_shape))
red_c <- crop(red, vect(clip_shape)) |> mask(vect(clip_shape))
dem_c <- crop(dem, vect(clip_shape)) |> mask(vect(clip_shape))

# Calculate NDVI
ndvi <- (nir_c - red_c) / (nir_c + red_c)

# Calculate slope (in percent)
slope <- terrain(dem_c, v = "slope", unit = "degrees")
slope_percent <- tan(slope * pi / 180) * 100

# Resample slope to match NDVI (so they can be added)
slope_resampled <- resample(slope_percent, ndvi)

# Combine with weights
combined <- 0.6 * ndvi + 0.4 * slope_resampled

# Reclassify using Natural Breaks (Fisher-Jenks)
vals <- values(combined, na.rm = TRUE)
fisher <- classIntervals(vals, n = 5, style = "fisher")
classified <- classify(combined, fisher$brks, include.lowest = TRUE)

# Assign category labels
categories <- c("Very Low", "Low", "Medium", "High", "Very High")
classified <- as.factor(classified)
levels(classified) <- data.frame(ID = 1:5, Category = categories)

# Reproject and resample to EPSG:32717 with 30m resolution
classified_utm <- st_warp(
  src = st_as_stars(classified),
  crs = 32717,
  cellsize = 30,
  method = "near"  # preserve class values
)

# Create coordinate grid (graticule) in EPSG:32717
tm_shape(classified_utm) +
  tm_raster(title = "Susceptibility", palette = "-RdYlGn", labels = categories) +
  tm_grid(
    crs = "EPSG:32717",
    n.x = 6,
    n.y = 6,
    xlabels = TRUE,
    ylabels = TRUE,
    labels.inside.frame = FALSE,
    col = "black",
    alpha = 0.4,
    labels.size = 0.8
  ) +
  tm_layout(
    main.title = "Landslide Susceptibility",
    legend.outside = TRUE,
    legend.outside.position = "right",
    frame = FALSE
  ) +
  tm_xlab("Easting (m, UTM Zone 17S)") +
  tm_ylab("Northing (m)")

# Export to GeoTIFF
output_path <- file.path(workspace, "landslide_30m.tif")
write_stars(classified_utm, output_path, overwrite = TRUE)

File: 12_MAP_ALGEBRA/12_MAP_ALGEBRA.py
import arcpy
from arcpy.sa import *
import os

# Activate the Spatial Analyst extension
arcpy.CheckOutExtension("Spatial")
arcpy.env.overwriteOutput = True

# Base directory (input rasters and output location)
workspace = r"C:\Users\alvarece\Desktop\MODELAMIENTO Y PROGRAMACION GIS\PRACTICAS\DATOS\RASTER"
arcpy.env.workspace = workspace

# Input rasters and shapefile
raster1 = os.path.join(workspace, "LC09_20240831_B5.tif")  # NIR band
raster2 = os.path.join(workspace, "LC09_20240831_B4.tif")  # Red band
raster3 = os.path.join(workspace, "DTM.tif")               # DEM
clip_shp = os.path.join(workspace, "QUITO.shp")            # Clip area shapefile

# Function to reproject raster to EPSG:4326 if needed
def reproject_if_needed(raster_path, name_in_memory):
    desc = arcpy.Describe(raster_path)
    sr = desc.spatialReference
    if sr.factoryCode != 4326:
        print(f"Reprojecting {os.path.basename(raster_path)} from EPSG:{sr.factoryCode} to EPSG:4326")
        return arcpy.management.ProjectRaster(raster_path, f"in_memory/{name_in_memory}", arcpy.SpatialReference(4326))
    else:
        print(f"{os.path.basename(raster_path)} is already in EPSG:4326")
        return raster_path

# Reproject rasters conditionally
r1_proj = reproject_if_needed(raster1, "r1_proj")
r2_proj = reproject_if_needed(raster2, "r2_proj")
r3_proj = reproject_if_needed(raster3, "r3_proj")

# Clip rasters to shapefile
clip_r1 = arcpy.management.Clip(r1_proj, "#", "in_memory/r1_clip", clip_shp, "0", "ClippingGeometry", "MAINTAIN_EXTENT")
clip_r2 = arcpy.management.Clip(r2_proj, "#", "in_memory/r2_clip", clip_shp, "0", "ClippingGeometry", "MAINTAIN_EXTENT")
clip_r3 = arcpy.management.Clip(r3_proj, "#", "in_memory/r3_clip", clip_shp, "0", "ClippingGeometry", "MAINTAIN_EXTENT")

# Calculate NDVI: (NIR - Red) / (NIR + Red)
ndvi = (Raster(clip_r1) - Raster(clip_r2)) / (Raster(clip_r1) + Raster(clip_r2))

# Calculate slope (in percent) from DEM
slope = Slope(Raster(clip_r3), output_measurement="PERCENT_RISE")

# Add NDVI and slope
sum_raster = 0.6*ndvi + 0.4*slope

# Classify the result into 5 classes using Natural Breaks (Jenks)
classified = Slice(sum_raster, 5, "NATURAL_BREAKS")

# Save final classified raster
classified_path = os.path.join(workspace, "final_classified.tif")
classified.save(classified_path)

print("✅ Process completed successfully!")
print("📁 Classified raster saved at:", classified_path)
